Qute - Validate DataNamespace expression if var has param declaration

michalvavrik opened the pull request:
Resolves: #20771

DataNamespace is currently not validated, this PR adds the validation if a parameter declaration is bound to the variable.

michalvavrik commented:
cc @mkouba

michalvavrik commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:
```diff
@@ -591,15 +591,18 @@ public String apply(String id) {
             for (Expression expression : templateAnalysis.expressions) {
-                if (expression.isLiteral()) {
+                // Make it possible to validate data namespace
+                Expression validationReadyExpression = expression.inValidationMode();
```
Indeed, much easier, thanks. I didn't exactly follow all your hints. If you still can see flaws/bugs there, just tell what should I test and I will write a test and find how to fix it myself (-> less time spend for you). I can't fully comprehend `validateNestedExpressions` but my tests are passing.

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:
```diff
@@ -749,6 +752,34 @@ static Match validateNestedExpressions(QuteConfig config, TemplateAnalysis templ
+                Expression.Part firstPart = expression.getParts().get(0);
+
+                String typeInfo = null;
+                if (checkedTemplate != null) {
```
I think that this branch is unnecessary because `Template#getParameterDeclarations()` should contain the final set of param declarations...

michalvavrik replied:
It didn't yesterday, I'll check again and get back to you.

mkouba replied:
It's actually a non-sense because the "implicit" param declarations are only available during the build phase.

In any case, your code should probably first iterate over the "explicit" param declarations defined in the template and then try the `checkedTemplate.bindings`.

michalvavrik replied:
I've switched the order as required and added validation for Message Bundle Data Namespace expressions (not sure how this will be taken ...).

mkouba commented on independent-projects/qute/core/src/main/java/io/quarkus/qute/Parser.java:
```diff
@@ -1044,14 +1052,16 @@ private List<TemplateNode> readLines(List<List<TemplateNode>> lines, List<Templa
-                    || node == COMMENT_NODE) {
+            } else if (node instanceof SectionNode
+                    || (node instanceof ParameterDeclarationNode
+                            && !dataNamespaceKeys.contains(((ParameterDeclarationNode) node).getKey()))
```
What's this for? I mean `isStandalone()` is used to detect [standalone lines](https://quarkus.io/guides/qute-reference#removing-standalone-lines-from-the-template) and I see no connection with validation of `data:` expressions...

michalvavrik replied:
we need parameter declarations for "data" namespace, but they are removed as standalone if there is no key associated with them (so we need to know there is an expression in data namespace associated with them).

mkouba replied:
Ah, you're right. But that's more like an oversight/bug. We need to fix this.

mkouba replied:
FYI https://github.com/quarkusio/quarkus/pull/26254

michalvavrik commented:
PR ready for next review round.

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/QuteProcessor.java:
```diff
@@ -55,6 +56,8 @@
 import org.jboss.jandex.TypeVariable;
 import org.jboss.logging.Logger;
 
+import com.google.common.collect.ImmutableMap;
```
Pls do not use guava in this module. You can use the immutable map from JDK (`Map.of()`, `Map.copyOf()`, etc.).

michalvavrik replied:
Sure, fixed, thanks

mkouba commented on extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/typesafe/DataNamespaceArrayValidationFailureTest.java:
```diff
@@ -0,0 +1,44 @@
+                    e = e.getCause();
+                }
+                assertNotNull(te);
+                assertTrue(te.getMessage().contains("Non-literal value used in bracket notation [i]"), te.getMessage());
```
This error is thrown by the parser, i.e. has no connection to the `data:` expression validation, or?

michalvavrik replied:
My bad, now I properly read how the array/collections validation is done, I corrected test. Thanks

mkouba commented on extensions/qute/deployment/src/main/java/io/quarkus/qute/deployment/TemplatesAnalysisBuildItem.java:
```diff
@@ -40,13 +42,30 @@ public static final class TemplateAnalysis {
+        /**
+         * Implicit parameter declarations extracted from Message Bundle methods used for the Data namespace validation.
+         */
+        public final Map<String, String> msgBundleMethodParamNameToType;
```
I believe that we should make the message bundle params part of the `TemplateAnalysis.parameterDeclarations` instead. And we could do the same for checked templates. So that `templateAnalysis.parameterDeclarations` would be the only source of declarations for [the branch above](https://github.com/quarkusio/quarkus/pull/26218/files#diff-409ae75ca5b45c330d6f027342c63933ff169ca45a313721bad52bc174e4694eR769-R802).

WDYT?

michalvavrik replied:
@mkouba I don't mind to do this. Before I do, I've some objections - implicit parameter declarations taken from methods are not TemplateNodes in a sense of current implementation - there is not origin (that has property lines etc.), there is no default. Also it has to be set before the template is parsed (to allow template param decl. to override it), so I'll have to iterate over once more then in current impl. Just confirm you prefer `parameterDeclarations` and I'll do that, it's easy.

mkouba replied:
> implicit parameter declarations taken from methods are not TemplateNodes in a sense of current implementation - there is not origin (that has property lines etc.), there is no default

That's ok. Just return `null` for both methods.

> Also it has to be set before the template is parsed (to allow template param decl. to override it), so I'll have to iterate over once more then in current impl. Just confirm you prefer parameterDeclarations and I'll do that, it's easy.

We can collect the implicit param declarations during the analysis (before the `ParserHook` is registered) and then merge it with `template.getParameterDeclarations()` (ofc the explicit declarations would take precedence).

mkouba commented on extensions/qute/deployment/src/test/java/io/quarkus/qute/deployment/typesafe/Item.java:
```diff
@@ -0,0 +1,23 @@
+
+    OtherItem[] otherItems;
+
+    public Item(String name, Set<OtherItem> otherItems) {
```
Just nitpicking... but why not `OtherItem... otherItems`?

michalvavrik replied:
I'm used to Sets rather than arrays. I pushed version with varargs as it looks better.

mkouba replied:
> I'm used to Sets rather than arrays.

Sure. It's just a little bit easier to use (and less allocations ;-).

mkouba reviewed (approved):
Looks good!

michalvavrik committed:
Qute - Validate DataNamespace expression if var has param declaration

Resolves: #20771

DataNamespace is currently not validated, this PR adds validation if parameter declaration is bound to the variable.

michalvavrik commented:
Failing test is not related to the changes.

mkouba commented:
Merged, thanks!
